src/Java.Interop.BootstrapTasks/Java.Interop.BootstrapTasks/JdkInfo.cs

```diff
@@ -14,74 +17,265 @@ public class JdkInfo : Task
+
+			WritePropertyFile (jarPath, javacPath, jdkJvmPath, includes);
+			WriteMakeFragmentFile (jarPath, javacPath, jdkJvmPath, includes);
+
+			Log.LogMessage (MessageImportance.Low, $"  [Output] {nameof (JavaHomePath)}: {JavaHomePath}");
 
 			return !Log.HasLoggedErrors;
 		}
 
-		string GetJavaSdkPath ()
+		Version GetMaxJdkVersion ()
+		{
+			if (string.IsNullOrEmpty (MaximumJdkVersion))
+				return null;
+			if (!MaximumJdkVersion.Contains (".")) {
+				MaximumJdkVersion += ".0";
+			}
+			return new Version (MaximumJdkVersion);
+		}
+
+		Version GetVersionFromPath (string path)
 		{
+			var m = VersionExtractor.Match (path);
+			if (!m.Success)
+				return null;
+			Version v;
+			if (!Version.TryParse (m.Groups ["version"].Value, out v)) {
+				return null;
+			}
+			return v;
+		}
+
+		void WritePropertyFile (string jarPath, string javacPath, string jdkJvmPath, IEnumerable<string> includes)
+		{
+			var msbuild = XNamespace.Get ("http://schemas.microsoft.com/developer/msbuild/2003");
+			var project = new XElement (msbuild + "Project",
+				new XElement (msbuild + "Choose",
+					new XElement (msbuild + "When", new XAttribute ("Condition", " '$(JdkJvmPath)' == '' "),
+						new XElement (msbuild + "PropertyGroup",
+							new XElement (msbuild + "JdkJvmPath", jdkJvmPath)),
+						new XElement (msbuild + "ItemGroup",
+							includes.Select (i => new XElement (msbuild + "JdkIncludePath", new XAttribute ("Include", i)))))),
+				new XElement (msbuild + "PropertyGroup",
+					new XElement (msbuild + "JavaCPath", new XAttribute ("Condition", " '$(JavaCPath)' == '' "),
+						javacPath),
+					new XElement (msbuild + "JarPath", new XAttribute ("Condition", " '$(JarPath)' == '' "),
+						jarPath)));
+			project.Save (PropertyFile.ItemSpec);
+		}
+
+		void WriteMakeFragmentFile (string jarPath, string javacPath, string jdkJvmPath, IEnumerable<string> includes)
+		{
+			using (var o = new StreamWriter (MakeFragmentFile.ItemSpec)) {
+				o.WriteLine ($"JI_JAR_PATH          := {jarPath}");
+				o.WriteLine ($"JI_JAVAC_PATH        := {javacPath}");
+				o.WriteLine ($"JI_JDK_INCLUDE_PATHS := {string.Join (" ", includes)}");
+				o.WriteLine ($"JI_JVM_PATH          := {jdkJvmPath}");
+			}
+		}
+
+		string GetJavaHomePathFromEnvironment ()
+		{
+			var java_home = Environment.GetEnvironmentVariable ("JAVA_HOME");
+			if (!string.IsNullOrEmpty (java_home))
+				return java_home;
+			return null;
+		}
+
+		string GetJavaHomePathFromMachine (Version maxVersion)
+		{
+			var java_homes  = GetJavaHomePathsFromDirectory (JdksRoot)
+				.Concat (GetJavaHomePathsFromJava ())
+				.Concat (GetJavaHomePathsFromWindowsRegistry ())
+				.Distinct ()
+				.Where (d => Directory.Exists (d))
+				.Select (jh => new {
+					Path    = jh,
+					Version = GetVersionFromPath (jh),
+				})
+				.Where (v => maxVersion == null ? true : v.Version <= maxVersion)
+				.OrderByDescending (v => v.Version)
+				.Select (v => v.Path)
+				.ToList ();
```

Comment:
While LINQ looks cool and is pretty concise, it may hurt performance - perhaps it would be better to do it the old fashioned way with loops?

Reply:
It looks neat but it isn't clear :) You have to follow it carefully to know what's going on (and to infer the types)

Reply:
I think the LINQ usage here is fine because it is only used as part of our build, not shipped to customers.

The alternative of using multiple loops to append to a `List<T>`, then sorting at the end--it is probably going to be just as verbose.